### PR TITLE
planner: add more test cases for non-prep plan cache

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: ref #36598

Problem Summary: planner: add more test cases for non-prep plan cache

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: executor/issuetest/executor_issue_test.go [hunk 1]
```diff
@@ -682,7 +682,7 @@ func TestIssue22231(t *testing.T) {
 	tk.MustExec("create table t_issue_22231(a datetime)")
 	tk.MustExec("insert into t_issue_22231 values('2020--05-20 01:22:12')")
 	tk.MustQuery("select * from t_issue_22231 where a >= '2020-05-13 00:00:00 00:00:00' and a <= '2020-05-28 23:59:59 00:00:00'").Check(testkit.Rows("2020-05-20 01:22:12"))
-	tk.MustQuery("show warnings").Check(testkit.Rows("Warning 1292 Truncated incorrect datetime value: '2020-05-13 00:00:00 00:00:00'", "Warning 1292 Truncated incorrect datetime value: '2020-05-28 23:59:59 00:00:00'"))
+	tk.MustQuery("show warnings").MultiCheckContain([]string{"Truncated incorrect datetime value: '2020-05-13 00:00:00 00:00:00'", "Truncated incorrect datetime value: '2020-05-28 23:59:59 00:00:00'"})
```
Comment (qw4990) [inline executor/issuetest/executor_issue_test.go:685]: Prepare to enable this switch by default in the future.

File: planner/core/plan_cache_param.go [hunk 2]
```diff
@@ -72,8 +72,8 @@ func (pr *paramReplacer) Enter(in ast.Node) (out ast.Node, skipChildren bool) {
 		}
 	case *driver.ValueExpr:
 		pr.params = append(pr.params, n)
-		param := ast.NewParamMarkerExpr(len(pr.params) - 1)      // offset is used as order in non-prepared plan cache.
-		param.(*driver.ParamMarkerExpr).Datum = *n.Datum.Clone() // init the ParamMakerExpr's Datum
+		param := ast.NewParamMarkerExpr(len(pr.params) - 1)  // offset is used as order in non-prepared plan cache.
+		n.Datum.Copy(&param.(*driver.ParamMarkerExpr).Datum) // init the ParamMakerExpr's Datum
```
Comment (qw4990) [inline planner/core/plan_cache_param.go:76]: Reduce one time of data copy.

File: planner/core/plan_cache_utils.go [hunk 3]
```diff
@@ -215,8 +216,8 @@ type planCacheKey struct {
 func (key *planCacheKey) Hash() []byte {
 	if len(key.hash) == 0 {
 		var (
-			dbBytes    = hack.Slice(key.database)
-			bufferSize = len(dbBytes) + 8*6 + 3*8
+			dbBytes    = hack.Slice(key.stmtText)
+			bufferSize = len(dbBytes) * 2
```
Comment (qw4990) [inline planner/core/plan_cache_utils.go:220]: Reduce times of buffer re-allocation.

**Review by hawkingrei — APPROVED** [review]

File: planner/core/plan_cache_utils.go [hunk 4]
```diff
@@ -215,8 +216,8 @@ type planCacheKey struct {
...
 		)
 		if key.hash == nil {
 			key.hash = make([]byte, 0, bufferSize)
```
Comment (hawkingrei) [inline planner/core/plan_cache_utils.go:223]: it may use ```sync.Pool``` to improve performance.

**Review by fzzf678 — APPROVED** [review]

File: planner/core/plan_cache_utils.go (hunk 4 above)
Comment (qw4990) [reply planner/core/plan_cache_utils.go:223]: Yes, I add a TODO comment there.

**Pull request merged**